Drain node action

jakewins opened the pull request:
Three commits:

- Refactor actions into smaller files (because Drain is big)
- Drain action
- Init feature to allow actions a one-time init; drain uses this to clean up crashes

jakewins committed:
Refactor actions into smaller files

To make space for the substantially more complex "drain"
action, so things don't get so cluttered.

jakewins committed:
Implement "node drain" as an action

This copies the drain behavior from kubectl, roughly:

- Cordon off node (and label it so we can find it if we crash)
- Drain node using evict API (so we respect PDBs etc)
- Cordon off node

It handles crashing by always looking for nodes labeled as
about-to-be-drained by the chaos monkey; if we find any node
with a label like that, we uncordon it before we continue.

Hence, if we crash mid-drain, we'll leave a cordoned-off halfway-drained
node behind, but we will uncordon it the next time we see it, and
before we try to cordon off any other node

jakewins committed:
Add init path to allow early cleanup of crashes

Drain currently cleans up after a crash, but only once
the drain action runs; this expands the Spec/Action interfaces
to have an "init" step that lets them do any one-time setup work.

That's then leveraged to just call the cleanup routine when the
drain action is initialized.

jgrobbel commented on chaoskube/action/node_drain.go:
```diff
@@ -0,0 +1,147 @@
+}
+
+func uncordonNode(client kubernetes.Interface, victim *v1.Node) error {
+	delete(victim.Labels, LabelMarmosetCordoned)
```
Should the label not be cleared last?

jakewins replied:
It's done atomically with the uncordon, so the node is uncordoned and unlabelled at the same time. That way we remove one possible in-between state we could end up in.

jgrobbel replied:
Ah its all  being applied through the update, for some reason thought `delete` was a method operating on the node. LGTM then 👍

jgrobbel reviewed (approved):
LGTM bar one question on order the removal of the label.

jakewins committed:
Add retries on out-of-date node

The drain op takes a long time, so the node is often
changed while that happens. Re-fetch the latest version
if we get a conflict error.
